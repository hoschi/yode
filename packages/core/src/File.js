import stampit from '@stamp/it'
import * as R from 'ramda'
import profiler from './profiler'
import { parseCode, getFunctionsFromAst, getAllContainerNodesRecursive, getAllChildrenNodesRecursive, printAst, addTextToNode, replaceNodeInAst, getNodeForFirstFoundType, isNodeDirty, getInnerMostFunctionNode, getMetaData } from './astBackedEditing'
import { getFunctionByText } from 'ast/compareFunctions'

let File = stampit().deepProps({
    id: undefined,
    syntaxError: undefined,
    text: undefined,
    unformattedText: undefined,
    ast: undefined,
    hasConnectedError: undefined,
    functions: undefined,
    functionsMap: undefined,
    functionsTreeRoot: undefined
}).methods({
    init({id, unformattedText}) {
        this.id = id

        // create formatted code
        let {error: syntaxError, ast} = parseCode(unformattedText)
        if (ast) {
            this.text = printAst(ast)
            // refresh ast from formatted code
            let {ast: astFormatted} = parseCode(this.text)
            this.ast = astFormatted
            let {functions, functionsTreeRoot} = getFunctionsFromAst(this.ast, this.id)
            functions.forEach((node) => {
                // create text for each function at start, because recast can't
                // keep all formatting when parsing code snippets instead of code
                // in file
                node.text = printAst(node)
                // and use this as editor text, so we don't start with "dirty" editors
                node.unformattedText = node.text
            })
            this.setFunctions(functions)
            this.functionsTreeRoot = functionsTreeRoot
            // set text in node
            this.ast.text = this.ast.unformattedText = this.text
        } else {
            this.text = unformattedText
            this.setFunctions([])
            this.functionsTreeRoot = {
                children: [],
                isRoot: true
            }
            this.syntaxError = syntaxError
            this.hasConnectedError = true
        }

        this.unformattedText = this.text
    },
    getMetaData() {
        return getMetaData(this)
    },
    setFunctions(functions) {
        this.functions = functions
        this.functionsMap = R.zipObj(functions.map(R.prop('customId')), functions)
    },
    updateFileAst(newText) {
        let stop = profiler.start('- file text update')
        let {error: syntaxError, ast} = parseCode(newText)
        if (syntaxError) {
            // broken code, wait for working code
            this.syntaxError = syntaxError
            this.unformattedText = newText
            this.hasConnectedError = true
            stop()
            return
        }
        this.syntaxError = undefined
        // error of the whole this is now gone, because this is the only editor which can be edited and change the error state
        this.hasConnectedError = false

        this.ast = ast
        this.text = printAst(ast)
        // add unformatted text for editor
        this.unformattedText = newText
        this.ast.text = newText
        this.ast.unformattedText = newText

        let {functions, functionsTreeRoot, removedFunctions} = getFunctionsFromAst(this.ast, this.id, this.functions)
        this.setFunctions(functions)
        this.functionsTreeRoot = functionsTreeRoot
        stop()
        return {
            removedFunctions
        }
    },
    updateFunctionAst(newText, oldFunction) {
        let newFunction

        let stop = profiler.start('- text update')
        let {error: syntaxError, ast} = parseCode(newText)
        if (syntaxError) {
            let oldFunctionWithError = {
                ...oldFunction,
                syntaxError,
                unformattedText: newText
            }
            // broken code, wait for working code
            let updatedFunctions = this.functions.map(f => {
                if (f.customId === oldFunction.customId) {
                    // update node with new created one
                    return oldFunctionWithError
                }
                return f
            })
            this.setFunctions(updatedFunctions)
            this.hasConnectedError = true
            stop()
            return {
                nodesToUpdate: [oldFunctionWithError],
                node: oldFunctionWithError
            }
        }

        // error of this file is gone
        oldFunction.syntaxError = undefined
        // error of the whole file is now gone, because this is the only editor which can be edited and change the error state
        this.hasConnectedError = false

        // ast parsing wraps text in other nodes, because it parses it
        // standalone and out of context of original text
        newFunction = getNodeForFirstFoundType(oldFunction.type, ast)
        if (!newFunction) {
            // not found, uh oh?!
            throw new Error('old node type not found')
        }

        // add formatted text for function node comparision
        addTextToNode(newFunction)
        // add unformatted text for editor
        newFunction.unformattedText = newText
        // save old id, because it is still the same function
        newFunction.customId = oldFunction.customId

        // replace ast of old function, with ast generated from newText, in file ast
        this.ast = replaceNodeInAst(newFunction, this.ast)

        // update file text with merged ast
        let fileText = printAst(this.ast)

        // update text if possible
        if (!isNodeDirty(this)) {
            // in sync, update
            this.text = fileText
            this.unformattedText = fileText
            this.ast.text = fileText
            this.ast.unformattedText = fileText
        }

        // get functions to compare, current changed one can't be matched
        let functionsToCompare = this.functions.filter(f => newFunction.customId !== f.customId)
        // update functions and port props from already known functions
        let {functions, functionsTreeRoot, removedFunctions} = getFunctionsFromAst(this.ast, this.id, functionsToCompare)
        this.setFunctions(functions)
        this.functionsTreeRoot = functionsTreeRoot

        // collect function nodes to update by walking up the linked list of
        // containing functions. This list is now up to date after running
        // getFunctionsFromAst
        let containerNodesToUpdate = getAllContainerNodesRecursive(newFunction.parentFunction)
        let childrenNodesToUpdate = getAllChildrenNodesRecursive(newFunction.children)
        let nodesToUpdate = containerNodesToUpdate.concat(childrenNodesToUpdate)

        stop()
        return {
            nodesToUpdate,
            node: newFunction,
            removedFunctions,
        }
    },
    findFunctionAroundCursor(node, cursor) {
        let innerFunctionNode = getInnerMostFunctionNode(node, cursor)
        if (!innerFunctionNode) {
            // can't find inner most function node, perhaps because code is not parsable
            return
        }
        let innerFunctionText = printAst(innerFunctionNode)
        let foundFunction = getFunctionByText(innerFunctionText, this.functions)
        return foundFunction
    }
})

export default File
