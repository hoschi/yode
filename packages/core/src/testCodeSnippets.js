export let letOnly = `
let only = 'simple'
`.trim()

// trailing space after name and before args parens
export let recastFormat = `
function add (a, b) {
    return a + b;
}
`.trim()

export let recastFormat_formatted = `
function add(a, b) {
    return a + b;
}
`.trim()

export let nestedFunctions = `
let myVar = 'some string'
function myFunc (a) {
    function innerFunc () {
        return 'hello'
    }
    return a + innerFunc()
}
`.trim()

export let nestedFunctions_removedOneFunc = `
let myVar = 'some string'
function myFunc (a) {
    return a
}
`.trim()

export let jsxComponentAndTest = `
import React from 'react'

let HelloComponent = ({name}) => {
    return <span>Hello&nbsp;{name}!</span>;
}

describe('HelloComponent', () => {
    let name = 'Walter';

    it('displays name', () => {
        function getTestName(a, b) {
            return a + b + name;
        }

        let myName = getTestName('I', 'Am', name);
        expect(<HelloComponent name={myName} />).contain('Hello IAmWalter!')
    })

    // function not recognized because not of a supported type
    it("doesn't work without name", function () {
        let myName = undefined;
        expect(<HelloComponent name={myName} />).contain('Hello !')
    });
})
`.trim()

export let jsxComponentAndTest_removedFuncs = `
import React from 'react'

let HelloComponent = ({name}) => {
    return <span>Hello&nbsp;{name}!</span>;
}
`.trim()

export let jsxComponentAndTest_withError = `
import React from 'react'

let HelloComponent = ({name}) => {
    return <span>Hello&nbsp;{name}!</span>;
}

describe('HelloComponent', () => {
    // ERROR HERE, sting not closed
    let name = 'Walter

    it('displays name', () => {
        function getTestName(a, b) {
            return a + b + name;
        }

        let myName = getTestName('I', 'Am', name);
        expect(<HelloComponent name={myName} />).contain('Hello IAmWalter!')
    })

    // function not recognized because not of a supported type
    it("doesn't work without name", function () {
        let myName = undefined;
        expect(<HelloComponent name={myName} />).contain('Hello !')
    });
})
`.trim()

export let jsxComponentAndTest_changedInnerMostFunction = `
import React from 'react'

let HelloComponent = ({name}) => {
    return <span>Hello&nbsp;{name}!</span>;
}

describe('HelloComponent', () => {
    let name = 'Walter';

    it('displays name', () => {
        function getTestName(a, b) {
            return a + b + name+'mytest';
        }

        let myName = getTestName('I', 'Am', name);
        expect(<HelloComponent name={myName} />).contain('Hello IAmWalter!')
    })

    // function not recognized because not of a supported type
    it("doesn't work without name", function () {
        let myName = undefined;
        expect(<HelloComponent name={myName} />).contain('Hello !')
    });
})
`.trim()

export let jsxComponentAndTest_changedOuterFunction = `
import React from 'react'

let HelloComponent = ({name}) => {
    return <span>Hello&nbsp;{name}!</span>;
}

describe('HelloComponent', () => {
    let name = 'Walterrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrr'

    it('displays name', () => {
        function getTestName(a, b) {
            return a + b + name;
        }

        let myName = getTestName('I', 'Am', name);
        expect(<HelloComponent name={myName} />).contain('Hello IAmWalter!')
    })

    // function not recognized because not of a supported type
    it("doesn't work without name", function () {
        let myName = undefined;
        expect(<HelloComponent name={myName} />).contain('Hello !')
    });
})
`.trim()

export let jsxClassesAndObjects = `
import React from 'react'

let getText = (name) => 'Hello ' + name;

class Welcome extends React.Component {
  // ClassMethod, not supported
  render() {
    let text = getText(this.props.name)
    return <h1>{text}</h1>;
  }
}

let MyObject = {
  name:'test',
  // ObjectMethod, not supported
  render() {
    let text = getText(this.name)
    return <span>{text}</span>;
  }
}
`.trim()
