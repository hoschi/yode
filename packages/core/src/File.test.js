let mockedId = 0
jest.mock('./customIdGenerator', () => {
    return {
        getNextId: () => ++mockedId
    }
})

import * as R from 'ramda'
import * as snip from './testCodeSnippets'
import File from './File'

beforeEach(() => {
    mockedId = 0
})

test('init - with valid text', function () {
    let unformattedText = snip.jsxComponentAndTest
    let s = File.create()
    s.init({
        id: 'myFile0',
        unformattedText
    })

    expect(s.id).toBe('myFile0')
    expect(s.text).toBe(unformattedText)
    expect(s.unformattedText).toBe(unformattedText)
    expect(s.functions).toHaveLength(4)
    expect(R.keys(s.functionsMap)).toEqual(['1', '2', '3', '4'])
    expect(s.functionsTreeRoot.isRoot).toBe(true)
    expect(s.functionsTreeRoot.children).toHaveLength(2)
    expect(s.syntaxError).toBeUndefined()
    expect(s.hasConnectedError).toBeUndefined()
})

test('init - with "unformatted" code', function () {
    let unformattedText = snip.recastFormat
    let s = File.create()
    s.init({
        id: 'myFile0',
        unformattedText
    })

    expect(s.functions).toHaveLength(1)
    expect(s.functions[0].text).toBe(snip.recastFormat_formatted)
})

test('init - with syntax error', function () {
    let unformattedText = snip.jsxComponentAndTest_withError
    let s = File.create()
    s.init({
        id: 'myFile0',
        unformattedText
    })

    expect(s.hasConnectedError).toBe(true)
    expect(s.syntaxError.toString()).toBe('SyntaxError: Unterminated string constant (9:15)')

    expect(s.id).toBe('myFile0')
    expect(s.text).toBe(unformattedText)
    expect(s.unformattedText).toBe(unformattedText)
    expect(s.functions).toHaveLength(0)
    expect(R.keys(s.functionsMap)).toHaveLength(0)
    expect(s.functionsTreeRoot.isRoot).toBe(true)
    expect(s.functionsTreeRoot.children).toHaveLength(0)
})

test('updateFunctionAst - nested function', function () {
    let unformattedText = snip.jsxComponentAndTest
    let s = File.create()
    s.init({
        id: 'myFile0',
        unformattedText
    })

    let oldFunction = s.functions[3]
    expect(oldFunction.text).toBe(`
function getTestName(a, b) {
    return a + b + name;
}
        `.trim())

    let newText = `
function getTestName(a, b) {
    return a + b + name+'mytest';
}
        `.trim()
    let result = s.updateFunctionAst(newText, oldFunction)
    expect(result.node.customId).toBe(oldFunction.customId)
    expect(result.node.text).toBe(newText)
    expect(result.node.unformattedText).toBe(newText)
    expect(result.node.syntaxError).toBeUndefined()
    expect(result.removedFunctions).toBeUndefined()
    expect(result.nodesToUpdate).toHaveLength(2)
    expect(result).toMatchSnapshot()
    expect(s.text).toBe(snip.jsxComponentAndTest_changedInnerMostFunction)

    newText = `
function getTestName(a, b) {
    return a + b + name + 'mytest' + 'another;
}
        `.trim()
    oldFunction = s.functions[3]
    result = s.updateFunctionAst(newText, oldFunction)
    // error is saved in node
    expect(result.node.syntaxError.toString()).toBe('SyntaxError: Unterminated string constant (2:37)')
    // AST text is still the old one
    expect(result.node.text).toBe(oldFunction.text)
    // unformatted text is that from editor
    expect(result.node.unformattedText).toBe(newText)
    // only update changed function to show error, AST not updated so connected functions not updated
    expect(result.nodesToUpdate).toHaveLength(1)
    // not files has error, node has error
    expect(s.syntaxError).toBeUndefined()
    // file is in connected error state though
    expect(s.hasConnectedError).toBe(true)

    expect(result.node.customId).toBe(oldFunction.customId)
    expect(result.node.unformattedText).toBe(newText)
    expect(result.removedFunctions).toBeUndefined()
    expect(result).toMatchSnapshot()

    // fixing the syntax error brings us to the state before the error
    newText = `
function getTestName(a, b) {
    return a + b + name + 'mytest' + 'another';
}
        `.trim()
    oldFunction = s.functions[3]
    result = s.updateFunctionAst(newText, oldFunction)
    expect(result.node.customId).toBe(oldFunction.customId)
    expect(result.node.text).toBe(newText)
    expect(result.node.unformattedText).toBe(newText)
    expect(result.node.syntaxError).toBeUndefined()
    expect(result.removedFunctions).toBeUndefined()
    expect(result.nodesToUpdate).toHaveLength(2)
    expect(s.hasConnectedError).toBe(false)
    expect(result).toMatchSnapshot()
})

test('updateFunctionAst - function with children', function () {
    let unformattedText = snip.jsxComponentAndTest
    let s = File.create()
    s.init({
        id: 'myFile0',
        unformattedText
    })

    let oldFunction = s.functions[1]
    expect(oldFunction.text).toBe(`
() => {
    let name = 'Walter';

    it('displays name', () => {
        function getTestName(a, b) {
            return a + b + name;
        }

        let myName = getTestName('I', 'Am', name);
        expect(<HelloComponent name={myName} />).contain('Hello IAmWalter!')
    })

    // function not recognized because not of a supported type
    it("doesn't work without name", function () {
        let myName = undefined;
        expect(<HelloComponent name={myName} />).contain('Hello !')
    });
}
`.trim())

    let newText = `
() => {
    let name = 'Walterrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrr'

    it('displays name', () => {
        function getTestName(a, b) {
            return a + b + name;
        }

        let myName = getTestName('I', 'Am', name);
        expect(<HelloComponent name={myName} />).contain('Hello IAmWalter!')
    })

    // function not recognized because not of a supported type
    it("doesn't work without name", function () {
        let myName = undefined;
        expect(<HelloComponent name={myName} />).contain('Hello !')
    });
}
`.trim()
    let result = s.updateFunctionAst(newText, oldFunction)
    expect(result.node.customId).toBe(oldFunction.customId)
    expect(result.node.text).toBe(newText)
    expect(result.node.unformattedText).toBe(newText)
    expect(result.node.syntaxError).toBeUndefined()
    expect(result.removedFunctions).toBeUndefined()
    expect(result.nodesToUpdate).toHaveLength(2)
    expect(result).toMatchSnapshot()
    expect(s.text).toBe(snip.jsxComponentAndTest_changedOuterFunction)

    newText = `
() => {
    let name = 'Walterrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrr' + 'foo

    it('displays name', () => {
        function getTestName(a, b) {
            return a + b + name;
        }

        let myName = getTestName('I', 'Am', name);
        expect(<HelloComponent name={myName} />).contain('Hello IAmWalter!')
    })

    // function not recognized because not of a supported type
    it("doesn't work without name", function () {
        let myName = undefined;
        expect(<HelloComponent name={myName} />).contain('Hello !')
    });
}
`.trim()
    oldFunction = s.functions[1]
    result = s.updateFunctionAst(newText, oldFunction)
    // error is saved in node
    expect(result.node.syntaxError.toString()).toBe('SyntaxError: Unterminated string constant (2:68)')
    // AST text is still the old one
    expect(result.node.text).toBe(oldFunction.text)
    // unformatted text is that from editor
    expect(result.node.unformattedText).toBe(newText)
    // only update changed function to show error, AST not updated so connected functions not updated
    expect(result.nodesToUpdate).toHaveLength(1)
    // not files has error, node has error
    expect(s.syntaxError).toBeUndefined()
    // file is in connected error state though
    expect(s.hasConnectedError).toBe(true)

    expect(result.node.customId).toBe(oldFunction.customId)
    expect(result.node.unformattedText).toBe(newText)
    expect(result.removedFunctions).toBeUndefined()
    expect(result).toMatchSnapshot()

    // fixing the syntax error brings us to the state before the error
    newText = `
() => {
    let name = 'Walterrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrr' + 'foo';

    it('displays name', () => {
        function getTestName(a, b) {
            return a + b + name;
        }

        let myName = getTestName('I', 'Am', name);
        expect(<HelloComponent name={myName} />).contain('Hello IAmWalter!')
    })

    // function not recognized because not of a supported type
    it("doesn't work without name", function () {
        let myName = undefined;
        expect(<HelloComponent name={myName} />).contain('Hello !')
    });
}
`.trim()
    oldFunction = s.functions[1]
    result = s.updateFunctionAst(newText, oldFunction)
    expect(result.node.customId).toBe(oldFunction.customId)
    expect(result.node.text).toBe(newText)
    expect(result.node.unformattedText).toBe(newText)
    expect(result.node.syntaxError).toBeUndefined()
    expect(result.removedFunctions).toBeUndefined()
    expect(result.nodesToUpdate).toHaveLength(2)
    expect(s.hasConnectedError).toBe(false)
    expect(result).toMatchSnapshot()
})
