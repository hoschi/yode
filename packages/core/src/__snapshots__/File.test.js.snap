// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateFunctionAst - function with children 1`] = `
Object {
  "node": ArrowFunctionExpression 1:0-18:1 {
    "children": "[
        {
          children: [
            {
              children: [],
              customId: 4,
            },
          ],
          customId: 3,
        },
      ]",
    "customId": 2,
    "fileId": "myFile0",
    "syntaxError": undefined,
    "text": "() => {
        let name = 'Walterrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrr'
    
        it('displays name', () => {
            function getTestName(a, b) {
                return a + b + name;
            }
    
            let myName = getTestName('I', 'Am', name);
            expect(<HelloComponent name={myName} />).contain('Hello IAmWalter!')
        })
    
        // function not recognized because not of a supported type
        it(\\"doesn't work without name\\", function () {
            let myName = undefined;
            expect(<HelloComponent name={myName} />).contain('Hello !')
        });
    }",
  },
  "nodesToUpdate": Array [
    ArrowFunctionExpression 4:24-11:5 {
      "children": "[
          {
            children: [],
            customId: 4,
          },
        ]",
      "customId": 3,
      "fileId": "myFile0",
      "syntaxError": undefined,
      "text": "() => {
          function getTestName(a, b) {
              return a + b + name;
          }
      
          let myName = getTestName('I', 'Am', name);
          expect(<HelloComponent name={myName} />).contain('Hello IAmWalter!')
      }",
    },
    FunctionDeclaration 5:8-7:9 {
      "children": "[]",
      "customId": 4,
      "fileId": "myFile0",
      "syntaxError": undefined,
      "text": "function getTestName(a, b) {
          return a + b + name;
      }",
    },
  ],
  "removedFunctions": undefined,
}
`;

exports[`updateFunctionAst - function with children 2`] = `
Object {
  "node": ArrowFunctionExpression 1:0-18:1 {
    "children": "[
        {
          children: [
            {
              children: [],
              customId: 4,
            },
          ],
          customId: 3,
        },
      ]",
    "customId": 2,
    "fileId": "myFile0",
    "syntaxError": [SyntaxError: Unterminated string constant (2:68)],
    "text": "() => {
        let name = 'Walterrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrr'
    
        it('displays name', () => {
            function getTestName(a, b) {
                return a + b + name;
            }
    
            let myName = getTestName('I', 'Am', name);
            expect(<HelloComponent name={myName} />).contain('Hello IAmWalter!')
        })
    
        // function not recognized because not of a supported type
        it(\\"doesn't work without name\\", function () {
            let myName = undefined;
            expect(<HelloComponent name={myName} />).contain('Hello !')
        });
    }",
    "unformattedText": "() => {
        let name = 'Walterrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrr' + 'foo
    
        it('displays name', () => {
            function getTestName(a, b) {
                return a + b + name;
            }
    
            let myName = getTestName('I', 'Am', name);
            expect(<HelloComponent name={myName} />).contain('Hello IAmWalter!')
        })
    
        // function not recognized because not of a supported type
        it(\\"doesn't work without name\\", function () {
            let myName = undefined;
            expect(<HelloComponent name={myName} />).contain('Hello !')
        });
    }",
  },
  "nodesToUpdate": Array [
    ArrowFunctionExpression 1:0-18:1 {
      "children": "[
          {
            children: [
              {
                children: [],
                customId: 4,
              },
            ],
            customId: 3,
          },
        ]",
      "customId": 2,
      "fileId": "myFile0",
      "syntaxError": [SyntaxError: Unterminated string constant (2:68)],
      "text": "() => {
          let name = 'Walterrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrr'
      
          it('displays name', () => {
              function getTestName(a, b) {
                  return a + b + name;
              }
      
              let myName = getTestName('I', 'Am', name);
              expect(<HelloComponent name={myName} />).contain('Hello IAmWalter!')
          })
      
          // function not recognized because not of a supported type
          it(\\"doesn't work without name\\", function () {
              let myName = undefined;
              expect(<HelloComponent name={myName} />).contain('Hello !')
          });
      }",
      "unformattedText": "() => {
          let name = 'Walterrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrr' + 'foo
      
          it('displays name', () => {
              function getTestName(a, b) {
                  return a + b + name;
              }
      
              let myName = getTestName('I', 'Am', name);
              expect(<HelloComponent name={myName} />).contain('Hello IAmWalter!')
          })
      
          // function not recognized because not of a supported type
          it(\\"doesn't work without name\\", function () {
              let myName = undefined;
              expect(<HelloComponent name={myName} />).contain('Hello !')
          });
      }",
    },
  ],
}
`;

exports[`updateFunctionAst - function with children 3`] = `
Object {
  "node": ArrowFunctionExpression 1:0-18:1 {
    "children": "[
        {
          children: [
            {
              children: [],
              customId: 4,
            },
          ],
          customId: 3,
        },
      ]",
    "customId": 2,
    "fileId": "myFile0",
    "syntaxError": undefined,
    "text": "() => {
        let name = 'Walterrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrr' + 'foo';
    
        it('displays name', () => {
            function getTestName(a, b) {
                return a + b + name;
            }
    
            let myName = getTestName('I', 'Am', name);
            expect(<HelloComponent name={myName} />).contain('Hello IAmWalter!')
        })
    
        // function not recognized because not of a supported type
        it(\\"doesn't work without name\\", function () {
            let myName = undefined;
            expect(<HelloComponent name={myName} />).contain('Hello !')
        });
    }",
  },
  "nodesToUpdate": Array [
    ArrowFunctionExpression 4:24-11:5 {
      "children": "[
          {
            children: [],
            customId: 4,
          },
        ]",
      "customId": 3,
      "fileId": "myFile0",
      "syntaxError": undefined,
      "text": "() => {
          function getTestName(a, b) {
              return a + b + name;
          }
      
          let myName = getTestName('I', 'Am', name);
          expect(<HelloComponent name={myName} />).contain('Hello IAmWalter!')
      }",
    },
    FunctionDeclaration 5:8-7:9 {
      "children": "[]",
      "customId": 4,
      "fileId": "myFile0",
      "syntaxError": undefined,
      "text": "function getTestName(a, b) {
          return a + b + name;
      }",
    },
  ],
  "removedFunctions": undefined,
}
`;

exports[`updateFunctionAst - nested function 1`] = `
Object {
  "node": FunctionDeclaration 1:0-3:1 {
    "children": "[]",
    "customId": 4,
    "fileId": "myFile0",
    "syntaxError": undefined,
    "text": "function getTestName(a, b) {
        return a + b + name+'mytest';
    }",
  },
  "nodesToUpdate": Array [
    ArrowFunctionExpression 10:24-17:5 {
      "children": "[
          {
            children: [],
            customId: 4,
          },
        ]",
      "customId": 3,
      "fileId": "myFile0",
      "syntaxError": undefined,
      "text": "() => {
          function getTestName(a, b) {
              return a + b + name+'mytest';
          }
      
          let myName = getTestName('I', 'Am', name);
          expect(<HelloComponent name={myName} />).contain('Hello IAmWalter!')
      }",
    },
    ArrowFunctionExpression 7:27-24:1 {
      "children": "[
          {
            children: [
              {
                children: [],
                customId: 4,
              },
            ],
            customId: 3,
          },
        ]",
      "customId": 2,
      "fileId": "myFile0",
      "syntaxError": undefined,
      "text": "() => {
          let name = 'Walter';
      
          it('displays name', () => {
              function getTestName(a, b) {
                  return a + b + name+'mytest';
              }
      
              let myName = getTestName('I', 'Am', name);
              expect(<HelloComponent name={myName} />).contain('Hello IAmWalter!')
          })
      
          // function not recognized because not of a supported type
          it(\\"doesn't work without name\\", function () {
              let myName = undefined;
              expect(<HelloComponent name={myName} />).contain('Hello !')
          });
      }",
    },
  ],
  "removedFunctions": undefined,
}
`;

exports[`updateFunctionAst - nested function 2`] = `
Object {
  "node": FunctionDeclaration 1:0-3:1 {
    "children": "[]",
    "customId": 4,
    "fileId": "myFile0",
    "syntaxError": [SyntaxError: Unterminated string constant (2:37)],
    "text": "function getTestName(a, b) {
        return a + b + name+'mytest';
    }",
    "unformattedText": "function getTestName(a, b) {
        return a + b + name + 'mytest' + 'another;
    }",
  },
  "nodesToUpdate": Array [
    FunctionDeclaration 1:0-3:1 {
      "children": "[]",
      "customId": 4,
      "fileId": "myFile0",
      "syntaxError": [SyntaxError: Unterminated string constant (2:37)],
      "text": "function getTestName(a, b) {
          return a + b + name+'mytest';
      }",
      "unformattedText": "function getTestName(a, b) {
          return a + b + name + 'mytest' + 'another;
      }",
    },
  ],
}
`;

exports[`updateFunctionAst - nested function 3`] = `
Object {
  "node": FunctionDeclaration 1:0-3:1 {
    "children": "[]",
    "customId": 4,
    "fileId": "myFile0",
    "syntaxError": undefined,
    "text": "function getTestName(a, b) {
        return a + b + name + 'mytest' + 'another';
    }",
  },
  "nodesToUpdate": Array [
    ArrowFunctionExpression 10:24-17:5 {
      "children": "[
          {
            children: [],
            customId: 4,
          },
        ]",
      "customId": 3,
      "fileId": "myFile0",
      "syntaxError": undefined,
      "text": "() => {
          function getTestName(a, b) {
              return a + b + name + 'mytest' + 'another';
          }
      
          let myName = getTestName('I', 'Am', name);
          expect(<HelloComponent name={myName} />).contain('Hello IAmWalter!')
      }",
    },
    ArrowFunctionExpression 7:27-24:1 {
      "children": "[
          {
            children: [
              {
                children: [],
                customId: 4,
              },
            ],
            customId: 3,
          },
        ]",
      "customId": 2,
      "fileId": "myFile0",
      "syntaxError": undefined,
      "text": "() => {
          let name = 'Walter';
      
          it('displays name', () => {
              function getTestName(a, b) {
                  return a + b + name + 'mytest' + 'another';
              }
      
              let myName = getTestName('I', 'Am', name);
              expect(<HelloComponent name={myName} />).contain('Hello IAmWalter!')
          })
      
          // function not recognized because not of a supported type
          it(\\"doesn't work without name\\", function () {
              let myName = undefined;
              expect(<HelloComponent name={myName} />).contain('Hello !')
          });
      }",
    },
  ],
  "removedFunctions": undefined,
}
`;
